Move conformal maps to library

AdelekeBankole opened the pull request.

jeremylt committed:
advection - initial advection operators and transplanted quadrature

Co-authored-by: Adeleke Bankole <AdelekeBankole@colorado.edu>
Co-authored-by: Jed Brown <jed@jedbrown.org>

jedbrown committed:
Enable mappings via keyword arguments

jedbrown committed:
ex010_advection: use LinRange, positive frequencies, axis scaling

jedbrown committed:
Add transform for gallery operators

jedbrown committed:
Add Kosloff-Tal-Ezer and Hale-Trefethen strip transforms

jedbrown committed:
Style fixes

AdelekeBankole committed:
Move conformal maps to library

jedbrown commented on src/Basis/Constructors.jl:
```diff
@@ -17,7 +34,74 @@ function sausage(d = 9)
+- `α`: polynomial degree of truncated Taylor series expansion of arcsin(s).
+
+# Returns:
+The Kosloff and Tal-Ezer map conformal map derived from the inverse sine function.
```
```suggestion
The Kosloff and Tal-Ezer conformal map derived from the inverse sine function.
```

jedbrown commented on src/Basis/Constructors.jl:
````diff
@@ -4,11 +4,28 @@
+# Example:
+```jldoctest
+# sausage conformal map
+g, derivative(g) = LFAToolkit.sausage(9);
````
Is this the right syntax?

AdelekeBankole replied:
Uhm, derivative(g) should be written as g_prime. With this it will be consistent with the other two conformal maps

jedbrown reviewed (commented):
I think you want to target this PR to `main` rather than Jeremy's `transplanted_quadrature` branch. That'll likely require some squashing of commits.

jedbrown commented on src/Basis/Constructors.jl:
```diff
@@ -2,28 +2,151 @@
 # finite element bases
 # ------------------------------------------------------------------------------
 
+using Polynomials
```
I think this is no longer used.

AdelekeBankole replied:
from checking today, I think we still need it in here.

jedbrown commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,54 @@
+
+# setup
+mesh = Mesh1D(1.0)
+P = 11;
```
No need to be very high order in the demo. Better to run fast and be easy to interpret.
```suggestion
P = 4;
```

jedbrown commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,54 @@
+basis = TensorH1LagrangeBasis(P, Q, 1, 1, collocatedquadrature = collocate, mapping = mapping)
+
+# frequency set up
+numbersteps = 250
```
The standard amount should be enough.
```suggestion
numbersteps = 100
```

jedbrown commented on examples/ex011_advection_mapped.jl:
```diff
@@ -0,0 +1,43 @@
+# Trefethen (2008) SIAM J. NUMER. ANAL. Vol. 46, No. 2
+# ---------------------------------------------------------------
+using LFAToolkit
+using LinearAlgebra
+using Plots
+
+# setup
+mesh = Mesh1D(1.0)
+P = 3
+Q = P
+basis = TensorH1LagrangeBasis(P, Q, 1, 1)
+
+# associated phase speed
+h0 = 1.0
+g = 9.81
+U = sqrt(g * h0)
+
+# weak form
+function advectionmappedweakform(u::Array{Float64}, w::Array{Float64})
+    dv = U * u * w[1]
+    return [dv]
+end
+
+# mapped advection operator
+# multiply differentiation basis matrix by inverse of g' 
+# multiply quadrature weights by g'
+inputs = [
+    OperatorField(basis, [EvaluationMode.interpolation]),
+    OperatorField(basis, [EvaluationMode.quadratureweights]),
+]
+outputs = [OperatorField(basis, [EvaluationMode.gradient])]
+advectionmapped = Operator(advectionmappedweakform, mesh, inputs, outputs)
+mass = GalleryOperator("mass", P, Q, mesh)
+
+# compute operator symbols
+A = computesymbols(advectionmapped, [π])
+M = computesymbols(mass, [π])
+eigenvalues = real(eigvals(-M \ A))
+
+# ---------------------------------------------------------------
```
I think this file can be deleted.

AdelekeBankole replied:
It is deleted

jedbrown commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,54 @@
+# ---------------------------------------------------------------
+using LFAToolkit
+using LinearAlgebra
+using Polynomials
```
I think not used here either.
```suggestion
```

jedbrown reviewed (commented):
I reset the PR target. There are 7 commits that need to be squashed to avoid false starts. (Please ask if you need help with doing this or understanding why.) There's a failing style check (sort of weird line length reasons). When you push updates, the continuous integration will re-run and you'll want to confirm that it succeeds.

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
delete ex011_advection_mapped.jl

AdelekeBankole committed:
Merge branch 'transplanted_quadrature' of https://github.com/AdelekeBankole/LFAToolkit.jl into transplanted_quadrature

AdelekeBankole committed:
Fix g_prime syntax

AdelekeBankole committed:
Merge branch 'transplanted_quadrature' of https://github.com/AdelekeBankole/LFAToolkit.jl into transplanted_quadrature

AdelekeBankole committed:
advection problem

jeremylt commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,53 @@
+Q = P;
+collocate = false
+mapping = sausage(29)
+#mapping = kosloff_tal_ezer(0.98)
```
It would be good to clean up the dead code in this example

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -2,28 +2,152 @@
+# Example:
+```jldoctest
+# sausage conformal map
+g, derivative(g) = LFAToolkit.sausage(9);
````
these doctests should verify that the mapping make the correct output

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -2,28 +2,152 @@
+```julia
+sausage(d)
+```
+# Arguments:
````
The suggestion interface isn't working right here, but we have an extra line between the 3 backticks for the function name and the `# Arguments:` list (same with other new functions below)

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -47,7 +171,7 @@ trueweights = [
 ```
 """
-function gaussquadrature(q::Int)
+function gaussquadrature(q::Int; mapping = nothing)
````
Is there a type we can attach here? Maybe `mapping::function`? Just to help us use it right in the future

jedbrown replied:
good suggestion

AdelekeBankole replied:
@jeremylt what do you mean we do here ```mapping::function```?

jeremylt replied:
`q` has the type `Int`. The argument `mapping` should also have a type.

AdelekeBankole replied:
@jedbrown upon setting ```mapping::function``` in the library, where they are called. My toolkit would not even precompile anymore

jedbrown replied:
Push the commit and copy the error. It's probably a syntax mistake.

AdelekeBankole replied:
I pushed the commit 
```
julia> include("examples/ex010_advection.jl")
[ Info: Precompiling LFAToolkit [3f92b583-c0aa-4596-8bdf-f02f6c0a52df]
ERROR: LoadError: syntax: unexpected ")"
Stacktrace:
  [1] top-level scope
    @ ~/Projects/LFAToolkit.jl/src/Basis/Constructors.jl:139
  [2] include(mod::Module, _path::String)
    @ Base ./Base.jl:418
  [3] include(x::String)
    @ LFAToolkit ~/Projects/LFAToolkit.jl/src/LFAToolkit.jl:5
  [4] top-level scope
    @ ~/Projects/LFAToolkit.jl/src/LFAToolkit.jl:67
  [5] include
    @ ./Base.jl:418 [inlined]
  [6] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::String)
    @ Base ./loading.jl:1318
  [7] top-level scope
    @ none:1
  [8] eval
    @ ./boot.jl:373 [inlined]
  [9] eval(x::Expr)
    @ Base.MainInclude ./client.jl:453
 [10] top-level scope
    @ none:1
in expression starting at /home/adebanks/Projects/LFAToolkit.jl/src/Basis/Constructors.jl:139
in expression starting at /home/adebanks/Projects/LFAToolkit.jl/src/LFAToolkit.jl:5
ERROR: LoadError: Failed to precompile LFAToolkit [3f92b583-c0aa-4596-8bdf-f02f6c0a52df] to /home/adebanks/.julia/compiled/v1.7/LFAToolkit/jl_6kcYSv.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, ignore_loaded_modules::Bool)
   @ Base ./loading.jl:1466
 [3] compilecache(pkg::Base.PkgId, path::String)
   @ Base ./loading.jl:1410
 [4] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:1120
 [5] require(uuidkey::Base.PkgId)
   @ Base ./loading.jl:1013
 [6] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:997
 [7] include(fname::String)
   @ Base.MainInclude ./client.jl:451
 [8] top-level scope
   @ REPL[1]:1
in expression starting at /home/adebanks/Projects/LFAToolkit.jl/examples/ex010_advection.jl:11
```

jeremylt commented on docs/src/private/basis.md:
```diff
@@ -4,6 +4,10 @@
 LFAToolkit.gaussquadrature
 LFAToolkit.gausslobattoquadrature
+LFAToolkit.transformquadrature
+LFAToolkit.sausage
+LFAToolkit.kosloff_tal_ezer
+LFAToolkit.hale_trefethen_strip
```
Two notes here

1) Julia (tragically) prefers `koslofftalezer` over `kosloff_tal_ezer`, but I think this is a reasonable place to prefer underscores
https://docs.julialang.org/en/v1/manual/style-guide/#Use-naming-conventions-consistent-with-Julia-base/

2) Do we want to add the word transformation to make the purpose of these functions more clear? i.e. `sausage_transformation` and `kosloff_tal_ezer_transformation`?

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,151 @@
+{
```
ugh, I like these demos, but they are unplesant to review in the web view. I'll take a look from a local copy later

AdelekeBankole replied:
How about the jupyter notebook. Can you review it in the web view now?

jeremylt replied:
It's still messy in the PR web view, which is just how GitHub works. I'll take a look at it on my computer on Monday

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -2,28 +2,152 @@
+```
+"""
+function sausage(d)
+    c = zeros(d + 1)
+    c[2:2:end] = [1, cumprod(1:2:d-2) ./ cumprod(2:2:d-1)...] ./ (1:2:d)
+    c /= sum(c)
+    g = Polynomial(c)
+    g_prime = derivative(g)
+    g, g_prime
````
I suppose we have the choice here - do we want to keep the dependency on Polynomial or refactor to do this 'by hand'?

jedbrown replied:
It's trivial to make the coefficients for `g_prime` here, but these get evaluated as functions later and that seems like enough to justify using the library.

jeremylt replied:
oh, I see what's going on. Thanks

jeremylt commented on docs/src/public/basis.md:
```diff
@@ -3,6 +3,7 @@
 The finite element basis objects represent the discretization of the problem.
 
 There are convenience constructors for H1 tensor product bases on uniformly spaced, Gauss-Lobatto, or Gauss-Legendre points with Gauss-Lobatto or Gauss-Legendre quadrature.
+There are also nonpolynomial bases available derived from conformal maps such as sausage map, Kosloff and Tal-Ezer map and the Hale and Trefethen strip transformation 
```
```suggestion
There are also nonpolynomial bases available derived from conformal mappings applied to the bases on the Gauss-Lobatto points.
These mappings include the sausage map, Kosloff and Tal-Ezer map, and the Hale and Trefethen strip transformation.
```

jeremylt commented:
This is failing in CI because some of the doctests aren't formatted correctly and because `julia .Style.jl` needs to be run.

AdelekeBankole committed:
remove dead code and fix formatting

AdelekeBankole committed:
remove dead code

AdelekeBankole committed:
fix styles

AdelekeBankole committed:
fix styles

AdelekeBankole committed:
add transform to library

AdelekeBankole committed:
verify doctests

AdelekeBankole committed:
doc for advection operator

AdelekeBankole committed:
add advection.md in examples

AdelekeBankole committed:
update references

AdelekeBankole committed:
fix advectionoperator in gallery

AdelekeBankole committed:
some failing tests in transforms

AdelekeBankole committed:
add mapping

AdelekeBankole committed:
still some test passing problems

AdelekeBankole committed:
working on test passing issues

jedbrown commented on src/Basis/Constructors.jl:
````diff
@@ -2,27 +2,182 @@
+# Example:
+```jldoctest
+# generate transformed quadrature points, weights with choice of conformal map
+points = 5;
+mapping = haletrefethenstriptransformation(1.4);
+mpoints, mweights = LFAToolkit.transformquadrature(points, true, mapping);
````
You're literally applying the mapping to `points`, which is just the integer 5. You probably intended to apply it to an array of points between -1 and 1. This is the source of the error you're observing.

AdelekeBankole replied:
we have some errors with the ```mweights```, with ```haletrefethenstrip``` map, we have ```NaN``` at endpoints of the interval. For the ```sausage``` map, it does not give ```NaN``` but sum of weights exceed 2

AdelekeBankole committed:
some errors with mapped weights

jedbrown committed:
isapprox in tests when comparing to small values

jedbrown committed:
transformquadrature: fix arg documentation and doctest

The Hale-Trefethen and Kosloff-Tal-Ezer transformations do not
accurately integrate constants except in the limit. The sausage(9) map
is pretty good once you have at least 5 Gauss points.

AdelekeBankole committed:
Fix min and max eigenvalues

AdelekeBankole committed:
Fix accuracy test

AdelekeBankole committed:
Merge branch 'transplanted_quadrature' of https://github.com/AdelekeBankole/LFAToolkit.jl into transplanted_quadrature

jedbrown commented on docs/src/examples/advection.md:
````diff
@@ -0,0 +1,32 @@
+The advection problem is given by
+
+```math
+\dfrac{\partial \zeta}{\partial t} + U \dfrac{\partial \zeta}{\partial x} = 0
````
Convention in other examples is that $u$ is the independent variable.
```suggestion
\dfrac{\partial u}{\partial t} + c \dfrac{\partial u}{\partial x} = 0
```

jedbrown commented on docs/src/examples/advection.md:
````diff
@@ -0,0 +1,32 @@
+\int_\Omega v \left(\dfrac{\partial \zeta}{\partial t} \right) dV - \int_\Omega U \zeta \left( \dfrac{\partial v}{\partial x} \right) dV = 0, \forall v \in V
+```
+
+for an appropriate test space ``V \subseteq H_0^1 \left( \Omega \right)`` on the domain.
````
Usually $H_0^1$ would mean functions with zero trace on Dirichlet boundaries, but there are no such boundaries so $H^1$ is okay.
```suggestion
for an appropriate test space ``V \subseteq H^1 \left( \Omega \right)`` on the domain.
```

Technically, this is an abuse of notation because even $\cos x$ is not in $L^2([-\infty, \infty])$. This abuse is benign here.

jedbrown commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,32 @@
+In this weak formulation, boundary terms have been omitted, as they are not present on the infinite grid for Local Fourier Analysis.
+
+For understanding about nonpolynomial bases, see paper New quadrature formulas from conformal maps
+SIAM J. NUMER. ANAL. Vol. 46, No. 2 (2008) by following Hale and Trefethen (2008)
```
@jeremylt I take it juliadoc doesn't support bibtex. Anyway, I think linking the DOI is most useful in this context.
```suggestion
For understanding about nonpolynomial bases, see paper Hale and Trefethen (2008) New quadrature formulas from conformal maps. https://doi.org/10.1137/07068607X
```

jedbrown commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,32 @@
+
+### LFAToolkit code
+
+The advection operator is a classical test case to see dipersion spectrum inside LFAToolkit.
```
```suggestion
The advection operator is a classical test case to see dispersion spectrum inside LFAToolkit.
```
There are spell checking plugins for vscode, which you may want to activate.

jedbrown commented on docs/src/public/basis.md:
```diff
@@ -3,12 +3,13 @@
 The finite element basis objects represent the discretization of the problem.
 
 There are convenience constructors for H1 tensor product bases on uniformly spaced, Gauss-Lobatto, or Gauss-Legendre points with Gauss-Lobatto or Gauss-Legendre quadrature.
+There are also nonpolynomial bases available derived from conformal maps such as sausagetransformation, koslofftalezertransformation and the haletrefethenstriptransformation.
```
```suggestion
There are also nonpolynomial bases available derived from conformal maps such as `sausagetransformation`, `koslofftalezertransformation`, and `haletrefethenstriptransformation`.
```

jedbrown commented on docs/src/references.md:
```diff
@@ -4,4 +4,8 @@ A. Brandt, *Multi-level adaptive solutions to boundary-value problems*, Math. Co
\ No newline at end of file
+M. Gutknecht and S. Röllin,  *The Chebyshev iteration revisited*, Parallel Computing, 28 (2002), pp. 263-283.
+
+N. Hale and L. N. Trefethen, *New quadrature formulas from conformal maps*, SIAM J. NUMER. ANAL. Vol. 46, No. 2, (2008), pp. 930-948.
```
Ideally we'd link DOI here. It would be written like this.
```suggestion
N. Hale and L. N. Trefethen, [New quadrature formulas from conformal maps](https://doi.org/10.1137/07068607X), SIAM Journal on Numerical Analysis. Vol. 46, No. 2, (2008), pp. 930-948.
```
The convention above is to spell out journal names.

jedbrown commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,56 @@
+
+# weak form
+function advectionweakform(u::Array{Float64}, w::Array{Float64})
+    dv = U * u * w[1]
```
```suggestion
c = 1.0

# weak form
function advectionweakform(u::Array{Float64}, w::Array{Float64})
    dv = c * u * w[1]
```

jedbrown commented on src/Basis/Constructors.jl:
```diff
@@ -127,17 +299,18 @@ truepoints = [-1.0, -√(3/7), 0.0, √(3/7), 1.0];
 @assert truepoints ≈ quadraturepoints
 
 # generate Gauss-Legendre-Lobatto points and weights
-quadraturepoints, quadratureweights = LFAToolkit.gausslobattoquadrature(5, true);
+mapping = sausagetransformation(9);
+quadraturepoints, quadratureweights = LFAToolkit.gausslobattoquadrature(5, true, mapping=mapping)
 
 # verify
-trueweights = [1/10, 49/90, 32/45, 49/90, 1/10];
-@assert trueweights ≈ quadratureweights
+@assert quadraturepoints ≈ [-0.9999999999999999, -0.5418159129215785, 0.0, 0.5418159129215785, 0.9999999999999999]
+@assert quadratureweights ≈ [0.18690312494113656, 0.5445666710618016, 0.5400742149974571, 0.5445666710618016, 0.18690312494113656]
```
The comment is wrong (these are mapped points), and also, we shouldn't remove the check for unmapped quadrature weights. I would split this into separate tests, first with the standard weights, then with mapping.

jedbrown commented on src/Operator/Constructors.jl:
```diff
@@ -553,10 +693,96 @@ function diffusionoperator(basis::AbstractBasis, mesh::Mesh)
+    # setup
+    U = 1.0
+    function advectionweakform(u::Array{Float64}, w::Array{Float64})
+        dv = U * u * w[1]
```
I don't think this is correct for 2D advection. You should make the velocity into a vector, and I'd make it an optional argument to the function. Something like
```suggestion
        dv = wind * u * w[1]
```

jedbrown commented on src/Operator/Constructors.jl:
````diff
@@ -553,10 +693,96 @@ function diffusionoperator(basis::AbstractBasis, mesh::Mesh)
+    gradient
+```
+"""
+function advectionoperator(basis::AbstractBasis, mesh::Mesh)
+    # setup
+    U = 1.0
````
```suggestion
function advectionoperator(basis::AbstractBasis, mesh::Mesh, wind = [1, 1])
```

jedbrown reviewed (commented):
This is coming along. I just have minor consistency stuff. You can mark the threads resolved as you complete them.

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update docs/src/public/basis.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

jeremylt commented on docs/src/public/operatorgallery.md:
```diff
@@ -2,7 +2,7 @@
 
 The following PDEs are available in the operator gallery.
 
-### Scalar Mass and Diffusion Operator
```
Really, this line (and below) should say Operator**s**

AdelekeBankole committed:
Update docs/src/references.md

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -2,26 +2,179 @@
 ```
 
-Construct a Gauss-Legendre quadrature
+Construct a Gauss-Legendre quadrature with the option of applying Conformal maps
````
Should the C be lowercase?
```suggestion
Construct a Gauss-Legendre quadrature with the option of applying conformal maps
```

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -99,23 +271,23 @@ function gaussquadrature(q::Int)
 ```
 
-Construct a Gauss-Legendre-Lobatto quadrature
+Construct a Gauss-Legendre-Lobatto quadrature with the option of applying Conformal maps
````
same question as above
```suggestion
Construct a Gauss-Legendre-Lobatto quadrature with the option of applying conformal maps
```

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -2,26 +2,179 @@
 
 # Returns:
-- Gauss-Legendre quadrature points and weights
+- quadrature points, weights
```
```suggestion
- Gauss-Legendre quadrature points and weights, with conformal mapping applied to the points, if provided
```

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -99,23 +271,23 @@ function gaussquadrature(q::Int)
 
 # Returns:
-- Gauss-Legendre-Lobatto quadrature points or points and weights
+- Gauss-Legendre-Lobatto quadrature points or points, weights and mapped version
```
```suggestion
- Gauss-Legendre-Lobatto quadrature points or points and weights, with conformal mapping applied to the points, if provided
```

jeremylt commented on src/Basis/Constructors.jl:
````diff
@@ -2,26 +2,179 @@
 # Example:
 ```jldoctest
-# generate Gauss-Legendre points and weights
+# generate Gauss-Legendre points, weights
````
Why the removal of all these `and`? Am I missing something?

AdelekeBankole replied:
I will put them back.

AdelekeBankole committed:
Update src/Operator/Constructors.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Operator/Constructors.jl

Co-authored-by: Jed Brown <jed@jedbrown.org>

AdelekeBankole committed:
Fix grammar

AdelekeBankole committed:
add tests for standard weights

AdelekeBankole committed:
resolve paths

AdelekeBankole committed:
change zeta to u

jedbrown commented:
I don't know if this is ready for another round of review. There are still several unresolved conversations above. @AdelekeBankole Please resolve what you believe is resolved and let us know when it's ready.

AdelekeBankole commented:
Do I have to click the ```Resolve conversation``` button for the addressed issues?

jedbrown commented:
Yes, click once the conversation is resolved. If you're not sure, ask a follow-up question. If you resolve everything and want us to review, say that in a comment.

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,149 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# dependencies\n",
```
This portion was updated to
```julia
# dependencies
using Pkg
Pkg.activate(".")
Pkg.instantiate()
Pkg.develop(path="../..")
using Plots
using LFAToolkit
using LinearAlgebra
```

AdelekeBankole replied:
I have this  
```
using Pkg
Pkg.activate("./")
Pkg.instantiate()
Pkg.develop(path="./../../../LFAToolkit.jl")
using Plots
using LFAToolkit
using LinearAlgebra 
using Polynomials
```

Why are we not using ```Polynomials``` anymore? We will need that when invoking the ```sausagetransformation```

jeremylt replied:
We're using `Polynomials` inside of the library. I'm not sure we need it here. In any case, the change to the path certainly should be fixed

AdelekeBankole replied:
Ok, I will change the path right away

jeremylt replied:
The `Polynomials` module isn't needed it seems

jedbrown replied:
Yeah, we aren't calling `Polynomials` here, it's just used by the library.

jeremylt commented on docs/src/private/basis.md:
```diff
@@ -4,6 +4,10 @@
 LFAToolkit.gaussquadrature
 LFAToolkit.gausslobattoquadrature
+LFAToolkit.transformquadrature
+LFAToolkit.sausagetransformation
+LFAToolkit.koslofftalezertransformation
+LFAToolkit.haletrefethenstriptransformation
```
I think this was clearer with the underscores since these functions have names of people
```suggestion
LFAToolkit.sausage_transformation
LFAToolkit.kosloff_talezer_transformation
LFAToolkit.hale_trefethen_strip_transformation
```

AdelekeBankole replied:
I think I was earlier using "with underscores" then you said the Toolkit seems not to like underscores. Or am I misunderstanding you before?

jeremylt replied:
I said before that the Julia style guide prefers no underscores in function names, unless is significantly improves readability. I think the underscores significantly improves readability in this case.

AdelekeBankole replied:
ok

AdelekeBankole replied:
just to be clear/sure once again. I should return all the underscores here in the docs and inside the library?

jeremylt replied:
Yup, I think underscores per the suggestion above would make it clearer

AdelekeBankole replied:
Alright, thanks!

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,149 @@
+    "end\n",
+    "\n",
+    "# operator\n",
+    "advection = GalleryOperator(\"advection\", p + 1, p + 1, mesh)\n",
```
Some minor tidying:
```julia
# setup
p = 10
dimension = 2
mapping = haletrefethenstriptransformation(1.4)

mesh = []
if dimension == 1
   mesh = Mesh1D(1.0)
elseif dimension == 2
   mesh = Mesh2D(1.0, 1.0)
end

# operator
mass = GalleryOperator("mass", p + 1, p + 1, mesh, mapping = mapping)
advection = GalleryOperator("advection", p + 1, p + 1, mesh, mapping = mapping)
```

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,149 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# compute full operator symbols\n",
```
I tried to update the plotting to match the other notebooks, but the 2D plotting doesn't seem to work
```julia
# compute full operator symbols
numbersteps = 100
maxeigenvalue = 0
θ_min = -π
θ_max = θ_min + 2π
θ_range = LinRange(θ_min, θ_max, numbersteps)

# compute and plot smoothing factor
# -- 1D --
if dimension == 1
    # compute
    maxeigenvalues = zeros(numbersteps)
    mineigenvalues = zeros(numbersteps)
    for i in 1:numbersteps    
        θ = [θ_range[i]]
        A = computesymbols(advection, θ) * 2
        M = computesymbols(mass, θ)
        eigenvalues = hcat(sort(imag.(eigvals(-M\A)))...)'
        maxeigenvalues[i] = max(eigenvalues...)
        mineigenvalues[i] = min(eigenvalues...)
        maxeigenvalue = max(maxeigenvalue, maxeigenvalues[i])
    end

    # plot
    println("max eigenvalue: ", maximum(maxeigenvalues))
    xrange = θ_range/π
    plot(
        xrange,
        xlabel="θ/π",
        xtickfont=font(12, "Courier"),
        [maxeigenvalues, mineigenvalues, π*xrange],
        ytickfont=font(12, "Courier"),
        ylabel="spectral radius",
        linewidth=3,
        label=["Maximum λ" "Minimum λ" "θ"],
        title="Advection Operator Symbol Maximal Eigenvalues"
    )
    ymin = minimum(mineigenvalues)
    ylims!(minimum([0, ymin * 1.1]), maximum(maxeigenvalues) * 1.1)
# -- 2D --
elseif dimension == 2
    # compute
    maxeigenvalues = zeros(numbersteps, numbersteps)
    mineigenvalues = zeros(numbersteps, numbersteps)
    for i in 1:numbersteps, j in 1:numbersteps
        θ = [θ_range[i], θ_range[j]]
        A = computesymbols(advection, θ) * 2
        M = computesymbols(mass, θ)
        eigenvalues = hcat(sort(imag.(eigvals(-M\A)))...)'
        maxeigenvalues[i, j] = max(eigenvalues...)
        mineigenvalues[i, j] = max(eigenvalues...)
        maxeigenvalue = max(maxeigenvalue, maxeigenvalues[i, j])
    end

    # plot
    println("max eigenvalue: ", maximum(maxeigenvalues))
    xrange = θ_range/π
    plot1 = heatmap(
        xrange,
        xlabel="θ/π",
        xtickfont=font(12, "Courier"),
        xrange,
        ylabel="θ/π",
        maxeigenvalues,
        ytickfont=font(12, "Courier"),
        title="Advection Operator Symbol Maximum Eigenvalues",
        transpose=true,
        aspect_ratio=:equal
    )
    xlims!(θ_min/π, θ_max/π)
    ylims!(θ_min/π, θ_max/π)
    plot2 = heatmap(
        xrange,
        xlabel="θ/π",
        xtickfont=font(12, "Courier"),
        xrange,
        ylabel="θ/π",
        mineigenvalues,
        ytickfont=font(12, "Courier"),
        title="Advection Operator Symbol Minimum Eigevalues",
        transpose=true,
        aspect_ratio=:equal
    )
    xlims!(θ_min/π, θ_max/π)
    ylims!(θ_min/π, θ_max/π)
    plot!(plot1, plot2, layout = (2, 1), size = (700, 1400))
end
```
Edit: I made a couple of small fixes

AdelekeBankole replied:
Pls, the 2D plot is not yet fixed in this recent push. I will further think and look into it.

AdelekeBankole committed:
Update docs/src/private/basis.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
fix path and work on 2D plot

AdelekeBankole committed:
rewrite mappings for easy readability

AdelekeBankole committed:
Merge branch 'transplanted_quadrature' of https://github.com/AdelekeBankole/LFAToolkit.jl into transplanted_quadrature

jeremylt commented:
I think the open issues are

1) Fix the 2D Advection example

2) Add a type annotation to `mapping` in the function arguments

AdelekeBankole committed:
work on 2D advection plot

jedbrown committed:
Operator getelementmatrix: fix for nonsymmetric operators

jedbrown commented on src/Operator/Base.jl:
```diff
@@ -435,8 +442,8 @@ function getelementmatrix(operator::Operator)
                             D[
-                                (currentfieldin+j-1)*numberquadraturepoints+q,
                                 currentfieldout*numberquadraturepoints+q,
+                                (currentfieldin+j-1)*numberquadraturepoints+q,
```
@jeremylt Is it possible this was only tested on symmetric operators? This commit seemed like a lot, but I think it's all needed.

jeremylt replied:
It is not only possible, it is exactly what happened

jedbrown replied:
Okay, I haven't yet investigated the linear elasticity failure. I might have botched something to do with components or the test was formerly passing for the wrong reason.

jedbrown replied:
Maybe I have another mistake in this commit.
```diff
diff --git i/src/PC/Chebyshev.jl w/src/PC/Chebyshev.jl
index ec22fe23..5355956f 100644
--- i/src/PC/Chebyshev.jl
+++ w/src/PC/Chebyshev.jl
@@ -119,6 +119,7 @@ function getoperatordiagonalinverse(preconditioner::Chebyshev)
     # assemble if needed
     if !isdefined(preconditioner, :operatordiagonalinverse)
         # retrieve diagonal and invert
+        @show preconditioner.operator.diagonal
         diagonalinverse = preconditioner.operator.diagonal^-1
 
         # store
```
And then an error on the next line
```julia-repl
julia> include("examples/ex301_linear_elasticity.jl")
preconditioner.operator.diagonal = [143589.7435897434 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 273504.27350427327 0.0 0.0 0.0 0.0 0.0 0
.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 273504.27350427315 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0
 519658.1196581195 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 191452.99145299115 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 355555.5555555552 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 355555.55555555504 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 656410.2564102558 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0
 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0
.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.
0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0
 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0
.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.
0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0]
ERROR: LoadError: SingularException(9)
```

jedbrown replied:
Sure enough, the old code prior to my change
```
preconditioner.operator.diagonal = [143589.7435897434 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 273504.27350427327 0.0 0.0 0.0 0.0 0.0 0
.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 273504.27350427315 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0
 519658.1196581195 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 191452.99145299115 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 355555.5555555552 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 355555.55555555504 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 656410.2564102558 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0
 0.0 143589.74358974342 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 273504.27350427327 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 191452.99145299118 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 355555.5555555553 0.0 0.0 0.0
 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 273504.27350427327 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0
.0 0.0 0.0 519658.1196581194 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 355555.5555555551 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.
0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 656410.2564102559 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 143589.7435897434 0.0 0.0
0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 191452.99145299126 0.0 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.
0 0.0 0.0 0.0 273504.2735042732 0.0 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 355555.5555555552 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0
 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 273504.2735042732 0.0 0.0 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 355555.5555555553 0.0 0
.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 519658.1196581193 0.0; 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
0.0 0.0 0.0 0.0 656410.2564102559]
```

AdelekeBankole committed:
Type annotation to mapping

AdelekeBankole committed:
Merge branch 'transplanted_quadrature' of https://github.com/AdelekeBankole/LFAToolkit.jl into transplanted_quadrature

jedbrown commented on src/Basis/Constructors.jl:
````diff
@@ -136,7 +136,7 @@ wsum = sum(mweights);
 ```
 """
-function transformquadrature(points, weights = nothing, mapping = nothing)
+function transformquadrature(points, weights = nothing, mapping::function)
````
You just made it a required argument instead of an optional argument. You need
```julia
function transformquadrature(points, weights = nothing, mapping::Function = nothing)
```

AdelekeBankole replied:
ok, alright! Pls, what about ```Function``` or ```function```
Should it be ```F``` or ```f```?

jedbrown replied:
How could you find out in 10 seconds at a Julia terminal?

AdelekeBankole committed:
Fix type annotation to mapping

AdelekeBankole commented:
```
julia> include("examples/ex010_advection.jl")
ERROR: LoadError: TypeError: in keyword argument mapping, expected Function, got a value of type Tuple{LFAToolkit.var"#g#27"{Float64, LFAToolkit.var"#gu#25"{Float64, Float64, Float64}}, LFAToolkit.var"#gprime#28"{Float64, LFAToolkit.var"#gprimeu#26"{Float64, Float64, Float64}, Float64}}
Stacktrace:
 [1] top-level scope
   @ ~/Projects/LFAToolkit.jl/examples/ex010_advection.jl:20
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:451
 [3] top-level scope
   @ REPL[1]:1
in expression starting at /home/adebanks/Projects/LFAToolkit.jl/examples/ex010_advection.jl:20
```

jedbrown commented:
Ah, we're missing one thing. How is a mapping represented? I.e., what gets returned when you call `sausage_transformation(9)` or the like?

jedbrown committed:
Fix type annotation for mapping tuple

jedbrown committed:
computesymbolsoverrange: add support for reducing bandwidth

In the case of advection, the symbol at positive and negative
frequencies are complex conjugates of each other so we only need to plot one.

jedbrown committed:
demo010_advection: reduce to p=1 in 2D

jedbrown commented:
I pushed the type annotation and 2D plotting that I think makes sense. Please experiment with the `demo010_advection` notebook and let us know how you interpret it.

jedbrown commented on src/Basis/Constructors.jl:
```diff
@@ -36,8 +36,11 @@ function sausage_transformation(d)
+    pgprime = derivative(pg)
+    # Lower from Polynomial to Function
+    g(x) = pg(x)
+    gprime(x) = pgprime(x)
```
@jeremylt It seems we have to do something like this to make the `Tuple{Function, Function}` match because it seems Julia never coerces automatically.

jeremylt replied:
That makes sense, I think I've seen this before

jedbrown commented on src/Operator/Base.jl:
```diff
@@ -371,12 +373,17 @@ function getelementmatrix(operator::Operator)
             Btcurrent = []
             for mode in output.evaluationmodes
                 if mode == EvaluationMode.interpolation
-                    numberfields += output.basis.numbercomponents
+                    numberfields += 1
+                    numbernodeoutputs += output.basis.numbercomponents
+                    numberquadratureoutputs += output.basis.numbercomponents
                     Btcurrent =
                         Btcurrent == [] ? output.basis.interpolation :
                         [Btcurrent; output.basis.intepolation]
                 elseif mode == EvaluationMode.gradient
-                    numberfields += output.basis.dimension * output.basis.numbercomponents
+                    numberfields += output.basis.dimension
```
It seems my `numberfields` change here was wrong, but I don't understand why it has a different meaning here than for inputs (line 324, 331).

jedbrown replied:
I pushed a fixup (8a9f245b6979b4b642f2cfaf048cfad3705821ef) and tests pass locally now, but @jeremylt can you compare with the above lines and sign off on this bit, or maybe the input handling needs an update?

jeremylt replied:
I think this is good

jedbrown committed:
fixup! Operator getelementmatrix: fix for nonsymmetric operators

jeremylt commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,31 @@
+
+### LFAToolkit code
+
+The advection operator is a classical test case to see dispersion spectrum inside LFAToolkit.
```
`````suggestion
The advection operator is a classical test case to see dispersion spectrum inside LFAToolkit.

````@eval
using Markdown
Markdown.parse("""
```julia
$(read("../../../examples/ex010_advection.jl", String))
```
""")
````
`````

jeremylt commented on docs/src/public/operatorgallery.md:
```diff
@@ -2,13 +2,13 @@
 The following PDEs are available in the operator gallery.
 
-### Scalar Mass and Diffusion Operator
+### Scalar Mass, Diffusion and Advection Operators
```
```suggestion
### Scalar Mass, Diffusion, and Advection Operators
```

jeremylt commented on examples/ex010_advection.jl:
```diff
@@ -0,0 +1,56 @@
+end
+
+eigenvalues = hcat(advection_symbol.(θ)...)'
+# ---------------------------------------------------------------
```
```suggestion
eigenvalues = hcat(advection_symbol.(θ)...)'

# ---------------------------------------------------------------
```

jeremylt commented on examples/jupyter/Project.toml:
```diff
@@ -2,3 +2,4 @@
 LFAToolkit = "3f92b583-c0aa-4596-8bdf-f02f6c0a52df"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+Polynomials = "f27b6e38-b328-58d1-80ce-0feddd5e7a45"
```
Not needed here

jeremylt commented on Project.toml:
```diff
@@ -5,6 +5,7 @@ version = "0.5.0"
 
 [deps]
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Polynomials = "f27b6e38-b328-58d1-80ce-0feddd5e7a45"
```
I don't think this is needed here?

AdelekeBankole replied:
@jeremylt, which of the packages do you mean is not needed in the ```Project.toml``` files here and in the jupyter folder?

jeremylt replied:
Polynomials isn't directly used in any of our examples anymore, right?

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -2,22 +2,182 @@
+    c[2:2:end] = [1, cumprod(1:2:d-2) ./ cumprod(2:2:d-1)...] ./ (1:2:d)
+    c /= sum(c)
+    pg = Polynomial(c)
+    pgprime = derivative(pg)
+    # Lower from Polynomial to Function
+    g(x) = pg(x)
```
```suggestion
    pgprime = derivative(pg)

    # Lower from Polynomial to Function
    g(x) = pg(x)
```

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -2,22 +2,182 @@
+function hale_trefethen_strip_transformation(ρ)
+    τ = π / log(ρ)
+    d = 0.5 + 1 / (exp(τ * π) + 1)
+    π2 = π / 2
+    # Unscaled functions of u
+    gu(u) = log(1 + exp(-τ * (π2 + u))) - log(1 + exp(-τ * (π2 - u))) + d * τ * u
+    gprimeu(u) = 1 / (exp(τ * (π2 + u)) + 1) + 1 / (exp(τ * (π2 - u)) + 1) - d
+    # Normalizing factor and scaled functions of s
+    C = 1 / gu(π / 2)
```
```suggestion
    π2 = π / 2

    # Unscaled functions of u
    gu(u) = log(1 + exp(-τ * (π2 + u))) - log(1 + exp(-τ * (π2 - u))) + d * τ * u
    gprimeu(u) = 1 / (exp(τ * (π2 + u)) + 1) + 1 / (exp(τ * (π2 - u)) + 1) - d

    # Normalizing factor and scaled functions of s
    C = 1 / gu(π / 2)
```

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -324,8 +513,9 @@ TensorH1LagrangeBasis(
-    dimension,
+    dimension;
     collocatedquadrature = false,
+    mapping=nothing,
```
```suggestion
    mapping = nothing,
```

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -342,16 +532,17 @@ Tensor product basis on Gauss-Legendre-Lobatto points with Gauss-Legendre (defau
 - `collocatedquadrature = false`:  Gauss-Legendre or Gauss-Legendre-Lobatto quadrature points,
                                        default: false, Gauss-Legendre-Lobatto
 
+# Keyword Arguments:
```
This should be combined with the previous 'Keyword Arguments' block two lines above and align everything to the right of the `:` accordingly

jeremylt commented on src/Basis/Constructors.jl:
```diff
@@ -342,16 +532,17 @@ Tensor product basis on Gauss-Legendre-Lobatto points with Gauss-Legendre (defau
                                        default: false, Gauss-Legendre-Lobatto
 
+# Keyword Arguments:
+- `mapping=nothing`:  sausage_transformation, kosloff_talezer_transformation, hale_trefethen_strip_transformation
```
```suggestion
- `mapping = nothing`:  quadrature point mapping - sausage, Kosloff-Talezer, or Hale-Trefethen strip transformation
                                          default: nothing, no transformation
```

jeremylt commented on src/LFAToolkit.jl:
```diff
@@ -33,7 +33,11 @@ export TensorBasis,
+    sausage_transformation,
+    kosloff_talezer_transformation,
+    hale_trefethen_strip_transformation
 # operato oields
```
strange typo here (my fault in an old commit) on line 41 - Should be `operator fields`

jeremylt commented on src/Operator/Constructors.jl:
````diff
@@ -8,7 +8,8 @@
 """
 ```julia
-GalleryOperator(name, numbernodes1d, numberquadraturepoints1d, mesh)
+GalleryOperator(name, numbernodes1d, numberquadraturepoints1d, mesh,
+collocatedquadrature, mapping)
````
Not sure if the formatting will change in linting, but we should callout the keyword args like we did for the bases
```suggestion
GalleryOperator(
    name,
    numbernodes1d,
    numberquadraturepoints1d,
    mesh;
    collocatedquadrature = false,
    mapping = nothing
)
```

jeremylt commented on src/Operator/Constructors.jl:
````diff
@@ -553,10 +693,94 @@ function diffusionoperator(basis::AbstractBasis, mesh::Mesh)
 
+"""
+```julia
+advectionoperator(basis, mesh)
+```
+Convenience constructor for advection operator
+
+# Weak form:
+- ``\\int \\nabla v u``
+
+# Arguments:
+- `basis`: basis for all operator fields to use
+- `mesh`:  mesh for operator
````
`wind` is an undocumented argument

jeremylt commented on src/Utilities.jl:
```diff
@@ -24,6 +24,7 @@ Compute the eigenvalues and eigenvectors of the symbol matrix for an operator ov
                 default: nothing
 - `θ_min`:  Bottom of range of θ, shifts range to [θ_min, θ_min + 2π],
                 default: -π / 2
+- `θ_band`: θ_max = θ_min + θ_band
```
```suggestion
- `θ_band`: θ_max = θ_min + θ_band
                default: 2π
```

jeremylt commented on src/Utilities.jl:
````diff
@@ -8,7 +8,7 @@
 """
 ```julia
-computesymbolsoverrange(operator, θ; mass=nothing, θ_min=-π/2)
+computesymbolsoverrange(operator, numbersteps1d; mass=nothing, θ_min=-π/2, θ_band=2π)
````
```suggestion
computesymbolsoverrange(operator, numbersteps1d; mass = nothing, θ_min = -π/2, θ_band = 2π)
```

jeremylt commented on src/Utilities.jl:
```diff
@@ -101,11 +102,12 @@ function computesymbolsoverrange(
     numbersteps1d::Int;
     mass::Union{Operator,Nothing} = nothing,
     θ_min::Float64 = -π / 2,
+    θ_band::Float64 = 2π,
```
; if we want these to really be keyword arguments as mentioned above
```suggestion
    mass::Union{Operator,Nothing} = nothing;
    θ_min::Float64 = -π / 2,
    θ_band::Float64 = 2π,
```

jeremylt reviewed (changes requested):
Here's a batch of small style, commenting, doc, etc fixes

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update docs/src/public/operatorgallery.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update examples/ex010_advection.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Basis/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Operator/Constructors.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Utilities.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Utilities.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update src/Utilities.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

jeremylt commented on src/Utilities.jl:
```diff
@@ -99,13 +101,14 @@ eigenvalues = real(eigenvalues);
     operator::Operator,
     numbersteps1d::Int;
-    mass::Union{Operator,Nothing} = nothing,
+    mass::Union{Operator,Nothing} = nothing;
     θ_min::Float64 = -π / 2,
+    θ_band::Float64 = 2π,
```
oops, mass is a keyword arg too

```suggestion
    mass::Union{Operator,Nothing} = nothing,
    θ_min::Float64 = -π / 2,
    θ_band::Float64 = 2π,
```

AdelekeBankole committed:
Update src/Utilities.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

jeremylt commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,40 @@
+## Advection operator
+
+This is an example of an advection operator by using polynomial bases vs nonpolynomial bases
+The choices of nonpolynomial bases available are:
+mapping = sausage_transformation(9)
+mapping = kosloff_talezer_transformation(0.98)
+mapping = hale_trefethen_strip_transformation(1.4)
```
```suggestion
This is an example of the 1D scalar advection problem.
```

jeremylt replied:
woops, typo in the suggestion - should be good now

jeremylt commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,40 @@
+
+### LFAToolkit code
+
+The advection operator is a classical test case to see dispersion spectrum inside LFAToolkit.
```
```suggestion
The advection operator is a classical test case to see dispersion spectrum inside LFAToolkit.
Here we show the advection operator on a non-polynomial basis derived from the Hale-Trefethen strip transformation applied to a H1 Lagrange basis.

For understanding about nonpolynomial bases, see paper Hale and Trefethen (2008) New quadrature formulas from conformal maps. https://doi.org/10.1137/07068607X
```

jeremylt commented on docs/src/examples/advection.md:
```diff
@@ -0,0 +1,40 @@
+for an appropriate test space ``V \subseteq H^1 \left( \Omega \right)`` on the domain.
+In this weak formulation, boundary terms have been omitted, as they are not present on the infinite grid for Local Fourier Analysis.
+
+For understanding about nonpolynomial bases, see paper Hale and Trefethen (2008) New quadrature formulas from conformal maps. https://doi.org/10.1137/07068607X
+
```
```suggestion
```

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update docs/src/examples/advection.md

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
compute eigenvalues over range

AdelekeBankole committed:
represent mapping as tuple

AdelekeBankole committed:
fix typo

AdelekeBankole committed:
add wind speed to arguments

AdelekeBankole committed:
fix diverging branch

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,218 @@
+      " [10] serve_notebook(pipename::String, outputchannel_logger::Base.CoreLogging.SimpleLogger; crashreporting_pipename::String)\n",
+      "    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.6.24/scripts/packages/VSCodeServer/src/serve_notebook.jl:136\n",
+      " [11] top-level scope\n",
+      "    @ ~/.vscode/extensions/julialang.language-julia-1.6.24/scripts/notebook/notebook.jl:32\n",
+      " [12] include(mod::Module, _path::String)\n",
+      "    @ Base ./Base.jl:418\n",
+      " [13] exec_options(opts::Base.JLOptions)\n",
+      "    @ Base ./client.jl:292\n",
+      " [14] _start()\n",
+      "    @ Base ./client.jl:495"
+     ]
+    }
+   ],
+   "source": [
+    "# setup\n",
+    "p = 4\n",
+    "dimension = 1\n",
+=======
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# setup\n",
+    "p = 1\n",
+    "dimension = 2\n",
+>>>>>>> 48dc89b931d819c8a441fe77fbbf998b0a57a77b
+    "mapping = hale_trefethen_strip_transformation(1.4)\n",
+    "\n",
+    "mesh = []\n",
+    "if dimension == 1\n",
+    "   mesh = Mesh1D(1.0)\n",
+    "elseif dimension == 2\n",
+    "   mesh = Mesh2D(1.0, 1.0)\n",
+    "end\n",
+    "\n",
+    "# operator\n",
+    "mass = GalleryOperator(\"mass\", p + 1, p + 1, mesh, mapping = mapping)\n",
+    "advection = GalleryOperator(\"advection\", p + 1, p + 1, mesh, mapping = mapping)"
+   ]
+  },
```
1) merge conflicts

2) output should be cleared before committing the  jupyter notebooks

AdelekeBankole committed:
remove Polynomials

AdelekeBankole committed:
compute symbols over range

jedbrown commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -149,18 +97,10 @@
-    "    maxeigenvalues = reshape(maximum((-imag(eigenvalues)); dims=2), (numbersteps, numbersteps))\n",
-    "    mineigenvalues = reshape(minimum((-imag(eigenvalues)); dims=2), (numbersteps, numbersteps))\n",
-    "    \n",
->>>>>>> 48dc89b931d819c8a441fe77fbbf998b0a57a77b
```
Here you've removed the code that's necessary to plot the advection operator (which has imaginary eigenvalues).

AdelekeBankole committed:
return Polynomials back to top level

AdelekeBankole committed:
fix imaginary eigenvalues for 2D advection

AdelekeBankole committed:
minor fix in Keyword Arguments

jedbrown commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -9,9 +9,27 @@
+   "execution_count": 1,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\u001b[32m\u001b[1m  Activating\u001b[22m\u001b[39m project at `~/Projects/LFAToolkit.jl/examples/jupyter`\n",
+      "\u001b[32m\u001b[1mPrecompiling\u001b[22m\u001b[39m project...\n",
+      "\u001b[32m  ✓ \u001b[39mLFAToolkit\n",
+      "  1 dependency successfully precompiled in 2 seconds (126 already precompiled)\n",
+      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
+      "\u001b[32m\u001b[1m   Installed\u001b[22m\u001b[39m MutableArithmetics ─ v1.0.4\n",
+      "\u001b[32m\u001b[1m   Installed\u001b[22m\u001b[39m Polynomials ──────── v3.1.5\n",
+      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/Projects/LFAToolkit.jl/examples/jupyter/Project.toml`\n",
+      "\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m `~/Projects/LFAToolkit.jl/examples/jupyter/Manifest.toml`\n",
+      " \u001b[90m [d8a4904e] \u001b[39m\u001b[92m+ MutableArithmetics v1.0.4\u001b[39m\n",
+      " \u001b[90m [f27b6e38] \u001b[39m\u001b[92m+ Polynomials v3.1.5\u001b[39m\n"
+     ]
+    }
+   ],
```
Need to clear the output so we aren't versioning all the churn as each person runs it.
![image](https://user-images.githubusercontent.com/3303/180558353-a776524f-2446-4e16-b388-d0017b19a029.png)

AdelekeBankole committed:
cleared output in notebook

jedbrown commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,2033 @@
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArsAAAV4CAIAAAAv2oJBAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nOzdeXgURd4H8OrJTSBACEeIRAGVgCKgiIgiKCKCssLuqqAiHqu4q+u6r/eBu67rve+qq6svuiJRULxAvGAFBOSIQMIlZwgEkXCEOxJyd71/lLTdv85Up9Iz0z2Z7+fx8WG6q485MlNT8+1faZxzBgAAACAV8PoEAAAAIArE/fWvf/X6HNQUFxc/9NBD+/fv79Onj9fn8ou6urrbbrstPz//kksu8eQEfvrpp8LCwm3btlVWVrZu3VrTNE9Ow+f27du3cePGH3/8sba2tlmzZnFxcZE8ellZ2d13371hw4Y ...
+      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"700\" height=\"1400\" viewBox=\"0 0 2800 5600\">\n<defs>\n  <clipPath id=\"clip320\">\n    <rect x=\"0\" y=\"0\" width=\"2800\" height=\"5600\"/>\n  </clipPath>\n</defs>\n<path clip-path=\"url(#clip320)\" d=\"\nM0 5600 L2800 5600 L2800  ...
```
There's still output here that wasn't cleared.

AdelekeBankole committed:
clear output

jedbrown reviewed (approved):
Looks good to me now.

jeremylt commented on test/runtests.jl:
```diff
@@ -35,6 +35,17 @@ end
+# advection operator example with conformal maps
+# ------------------------------------------------------------------------------
+
+@testset "  ex010: advection example                        " begin
```
```suggestion
@testset "  ex010: advection example                   " begin
```

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,160 @@
+   "source": [
+    "# dependencies\n",
+    "using Pkg\n",
+    "Pkg.activate(\"./\")\n",
```
```suggestion
    "Pkg.activate(\".\")\n",
```

jeremylt commented on examples/jupyter/demo010_advection.ipynb:
```diff
@@ -0,0 +1,160 @@
+   "source": [
+    "# setup\n",
+    "p = 4\n",
+    "dimension = 2\n",
```
```suggestion
    "dimension = 1\n",
```

jeremylt reviewed (changes requested):
just a couple of small fixes

AdelekeBankole committed:
Update test/runtests.jl

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update examples/jupyter/demo010_advection.ipynb

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>

AdelekeBankole committed:
Update examples/jupyter/demo010_advection.ipynb

Co-authored-by: Jeremy L Thompson <jeremy@jeremylt.org>
